feat: Integrate Clerk backend auth and replace legacy JWT logic

mohamad-jaafar-monty opened the pull request:
- Replaced manual JWT decoding with Clerk’s backend SDK integration
- Introduced `AuthenticatedUserDTO` dataclass to cleanly map Clerk payloads
- Centralized token validation and payload extraction inside `auth.py`
- Added dependency-injected `CurrentUser` using Clerk SDK and HTTPBearer
- Updated `git_tokens.py` to scope returned tokens to authenticated users
- Added new module `system_messages.py` to centralize error strings
- Installed `clerk-backend-api` as project dependency

<!-- This is an auto-generated comment: release notes by coderabbit.ai -->
## Summary by CodeRabbit

- **New Features**
  - Enforced Clerk authentication for all protected routes, requiring valid Clerk-issued JWT tokens.
  - Added structured API responses for token endpoints, filtering tokens by authenticated user.
  - Introduced centralized logging with console and rotating file handlers for improved error tracking.

- **Bug Fixes**
  - Improved error handling and messaging for authentication failures and malformed authorization headers.

- **Documentation**
  - Updated README with comprehensive Clerk authentication setup, backend integration details, and testing instructions.

- **Tests**
  - Added extensive unit tests for authentication flows, including Clerk integration and error scenarios.
  - Introduced fixtures for simulating Clerk authentication states and generating test JWTs.

- **Chores**
  - Updated dependencies for Clerk and JWT handling.
  - Added pytest configuration to streamline test discovery and execution.
<!-- end of auto-generated comment: release notes by coderabbit.ai -->

Commit:
feat: integrate Clerk backend auth and replace legacy JWT logic

- Replaced manual JWT decoding with Clerk’s backend SDK integration
- Introduced `AuthenticatedUserDTO` dataclass to cleanly map Clerk payloads
- Centralized token validation and payload extraction inside `auth.py`
- Added dependency-injected `CurrentUser` using Clerk SDK and HTTPBearer
- Updated `git_tokens.py` to scope returned tokens to authenticated users
- Added new module `system_messages.py` to centralize error strings
- Installed `clerk-backend-api` as project dependency

Commit:
formatting: Applied black formatter

Applied black formatter on
- app/utils/auth.py
- app/routes/git_tokens.py

Commit:
fix: manually added the dependency PyJWT which clerk-backend-api dependency relies on, and commented conflicting dependency

PyJwt which is used by clerk-backend-api was conflicting with the jwt dependency, jwt can be safely removed because it was a temporary solution only.

Commit:
refactor: separate payload field validation from DTO construction

- Modified AuthenticatedUserDTO.from_clerk_payload to return (missing_fields, instance)
- Removed exception-raising logic from DTO class for better separation of concerns
- Moved missing field validation to get_current_user to enable context-aware logging
- Improved internal debug logging to reflect partial payload issues without crashing early

Commit:
test: Integrate Clerk authentication with DTO mapping and secure token filtering

Unit tests:
- Valid payload → mapped correctly to DTO
- Missing fields → triggers HTTPException
- Signed out Clerk state → returns 401
- Malformed auth scheme → returns 401
- Authenticated user → receives only their tokens
- Unauthenticated → receives 401 with proper message
- Supabase errors → returns 503 with fallback message
- Different user ID in Clerk vs data → tokens not exposed

Commit:
fix: Added pytest.ini

There is a behaviour where pytest was not recognizing conftest.py, the only way to get to recognize it at least for me, was via adding pytest.ini and pointing it to the tests directory.

mohamad-jaafar-monty commented on tests/conftest.py:
```diff
@@ -401,3 +402,59 @@ def mock_supabase_client():
+            "v": 2
+        }
+        if payload_overrides:
+            payload.update(payload_overrides)
```
This is part of the test, the unit test, not actual code, its purpose is to create a mock, so we don't need to take it from the environment.

mohamad-jaafar-monty commented on app/utils/auth.py:
```diff
@@ -1,97 +1,121 @@
+		return missing_payload_fields, cls(**mapped_fields)
+
+
+def get_current_user(
+		request_from_context: Request,
+		auth_header: HTTPAuthorizationCredentials = Depends(http_bearer_security_schema),
+) -> AuthenticatedUserDTO:
```
According to the documentation of FastAPI when trying to define the security schema, [this documentation](https://fastapi.tiangolo.com/reference/security/). That's the way to actually implement the schema, especially if the method being used acts as a "middleware" between Clerk and API endpoint.

mohamad-jaafar-monty commented on app/utils/auth.py:
```diff
@@ -1,97 +1,121 @@
+		
+		raise HTTPException(
+			status_code=status.HTTP_401_UNAUTHORIZED,
+			detail=INVALID_BEARER_TOKEN_SCHEMA,
+			headers={"WWW-Authenticate": "Bearer"},
+		)
```
The current INVALID_BEARER_TOKEN_SCHEMA already states that the token is malformed, moreover, you don't want to reveal all the exact details to the user, this should be done in the logs.

Commit:
Refactor: integrated logging, and fixed certain things reported by coderabbit and sonarQube

- Added basic logging, and integrated it
- Enhanced the Readme.md to include full clerk implementation.
- refactored things like TODO, unused fields ...

Hayat55 commented on README.md:
```diff
@@ -308,49 +308,105 @@ This behavior is part of Supabase's underlying infrastructure and not an issue w
+- Incoming request passes through `get_current_user()` in `auth.py`
+- Token is verified via Clerk's public key and API key
+- If valid:
+  - Extracts fields like `sub`, `email`, `name`
```
add the configuration that should be added to clerk to add these info in the bearer token

mohamad-jaafar-monty replied:
should be resolved now

Hayat55 commented on app/utils/auth.py:
```diff
@@ -1,97 +1,137 @@
+			exc_info=True
+		)
+		
+		raise HTTPException(
```
unify response format . Please use APIResponse class from app/utils/api_response.py

mohamad-jaafar-monty replied:
Should be resolved now

Hayat55 commented on app/utils/auth.py:
```diff
@@ -1,97 +1,137 @@
+			status_code=status.HTTP_401_UNAUTHORIZED,
+			detail={
+				"error": "not_authenticated",
+				"summary": "Access denied. Authentication failed.",
```
add All messages on app/utils/constants.py to avoid duplication later

mohamad-jaafar-monty replied:
Should be resolved now

Commit:
Refactor: unified response, and messages

- Unified the response.
- Moved common messages to constants.py folder instead of system_messages that I created earlier, cause they serve the same purpose.
- updated the readme to include how to customize payload coming from clerk.
